packages/react-server/core/renderMiddleware.js

```diff
@@ -67,85 +69,41 @@ module.exports = function(req, res, next, routes) {
 
 		// TODO? pull this context building into its own middleware
 		var context = new RequestContext.Builder()
-				.setRoutes(routes)
+				.setRoutes(req.app.get("reactServerRoutes"))
 				.setDefaultXhrHeadersFromRequest(req)
 				.create({
 					// TODO: context opts?
 				});
 
 		// Need this stuff in for logging.
 		context.setServerStash({ req, res, start, startHR });
-
 		context.setMobileDetect(new MobileDetect(req.get('user-agent')));
 
-		var navigateDfd = Q.defer();
-
-		// setup navigation handler (TODO: should we have a 'once' version?)
-		context.onNavigate( (err, page) => {
-
-			if (!navigateDfd.promise.isPending()) {
-				logger.error("Finished navigation after FAILSAFE_ROUTER_TIMEOUT", {
-					page: context.page,
-					path: req.path,
-				});
-				return;
-			}
-
-			// Success.
-			navigateDfd.resolve();
-
-
-			if (err) {
-				// The page can elect to proceed to render
-				// even with a non-2xx response.  If it
-				// _doesn't_ do so then we're done.
-				var done = !(page && page.getHasDocument());
-
-				if (err.status === 301 || err.status === 302 || err.status === 307) {
-					if (done){
-						// This adds a boilerplate body.
-						res.redirect(err.status, err.redirectUrl);
-					} else {
-						// This expects our page to
-						// render a body.  Hope they
-						// know what they're doing.
-						res.set('Location', err.redirectUrl);
-					}
-				} else if (done) {
-					if (err.status === 404) {
-						next();
-					} else {
-						next(err);
-					}
-				}
-				if (done) {
-					logger.log("onNavigate received a non-2xx HTTP code", err);
-					handleResponseComplete(req, res, context, start, page);
-					return;
-				}
-			}
-			renderPage(req, res, context, start, page);
-
-		});
-
+		let timeoutDefer = Q.defer();
 
-		const timeout = setTimeout(navigateDfd.reject, FAILSAFE_ROUTER_TIMEOUT);
-
-		// Don't leave dead timers hanging around.
-		navigateDfd.promise.then(() => clearTimeout(timeout));
-
-		// If we fail to navigate, we'll throw a 500 and move on.
-		navigateDfd.promise.catch(() => {
-			logger.error("Failed to navigate after FAILSAFE_ROUTER_TIMEOUT", {
-				page: context.navigator.getCurrentRoute().name,
+		const timeout = setTimeout(() => {
+			let metaData = {
+				page: context.page,
 				path: req.path,
-			});
-			handleResponseComplete(req, res, context, start, context.page);
-			next({status: 500});
+			};
+			timeoutDefer.reject(new HttpError(`Navigation timed out after ${FAILSAFE_ROUTER_TIMEOUT}ms`, metaData));
+		}, FAILSAFE_ROUTER_TIMEOUT);
+
+		// setup navigation handler (TODO: should we have a 'once' version?)
+		Q.race([
+			context.onNavigatePromise(),
+			timeoutDefer.promise,
+		])
+		.tap(() => clearTimeout(timeout)) //Clear timeout if promise succeeds - not strictly neccessary
+		.then((page) => renderPage(req, res, context, page))
+		.then(() => endResponse(req, res))
+		.catch(next)
+		.finally(() => {
+			clearTimeout(timeout);
+			handleResponseComplete(req, res, context);
```

Comment:
This looks like it switches the order of execution between `endResponse` and `handleResponseComplete`.

Trying to think if that could cause any problems... 🤔